projects/sandbox/train/train/validation.py

```diff
@@ -191,32 +196,69 @@ def postprocess(self, preds: torch.Tensor) -> torch.Tensor:
         )
         return preds[0, 0]
 
-    def inject(self, X: torch.Tensor) -> torch.Tensor:
+    def threshold_snrs(self, waveforms: torch.Tensor, psds: torch.Tensor):
+        num_freqs = waveforms.shape[-1] // 2 + 1
+        if num_freqs != psds.shape[-1]:
+            psds = torch.nn.functional.interpolate(psds, (num_freqs,))
+
+        mask = torch.linspace(0, self.sample_rate / 2, num_freqs)
+        mask = mask >= self.highpass
+        mask = mask.to(waveforms.device)
+
+        snrs = gw.compute_network_snr(waveforms, psds, self.sample_rate, mask)
+        target_snrs = snrs.clamp(self.snr_thresh, 1000)
+        weights = target_snrs / snrs
+        return waveforms * weights.view(-1, 1, 1)
+
+    def inject(self, X: torch.Tensor, psds: torch.Tensor) -> torch.Tensor:
+        # downsample the background batch to give us
+        # the desired stride between injections
         X = X[:: self._injection_step]
+        psds = psds[:: self._injection_step]
+
+        # grab the next batch of injections and possibly
+        # reduce the background batch size to match it
+        # if we're at the end. Increment our injection
+        # index counter
         start = self._injection_idx
         stop = start + len(X)
-        waveforms = self.waveforms[start:stop]
-
-        start = waveforms.shape[-1] // 2 - self.kernel_size // 2
-        stop = start + self.kernel_size
+        waveforms = self.waveforms[start:stop].to(X.device)
+        psds = psds[: len(waveforms)]
+        X = X[: len(waveforms)]
+        self._injection_idx += len(waveforms)
+
+        # threshold the SNRs of the injections to the desired value.
+        waveforms = self.threshold_snrs(waveforms, psds)
+
+        # now cut out a window symmetrically about the
+        # coalescence time and inject it into the background
+        kernel_size = X.shape[-1]
+        start = waveforms.shape[-1] // 2 - kernel_size // 2
```

Comment:
We're always putting the coalescence point in the middle now? Will that be a temporary change?